<a href="https://colab.research.google.com/github/hellog2n/Example-Tensorflow-Python/blob/main/StyleGAN2-test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# StyleGAN2 Code Clone 하기
NVIDIA에서 제공하는 StyleGAN2 코드를 로드합니다.

In [1]:
!git clone https://github.com/NVlabs/stylegan2

Cloning into 'stylegan2'...
remote: Enumerating objects: 138, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 138 (delta 4), reused 10 (delta 4), pack-reused 127
Receiving objects: 100% (138/138), 599.47 KiB | 6.38 MiB/s, done.
Resolving deltas: 100% (62/62), done.


현재 경로를 클론한 stylegan2 폴더로 이동합니다.

In [6]:
%cd stylegan2

/content/stylegan2


---

# Tensorflow Version 변경
tensorflow version을 1.14로 변경합니다.
이 이유는 공식 홈페이지 소스코드는 tensorflow 1.14 ~ 1.15을 지원하기 때문입니다.

In [ ]:
!pip install tensorflow==1.14

In [ ]:
!pip install tensorflow-gpu==1.14

In [15]:
import tensorflow as tf
print(tf.__version__)

1.14.0


stylegan2 코드에서 제공하고 있는 GPU, CPU 확인을 위한 파일을 실행해서
현재 내가 CPU, GPU가 사용 가능한지 확인해봅니다.

```
CPU says hello.
GPU says hello.
```
가 뜨면 잘 작동한다는 것을 알 수 있습니다.



In [4]:
!nvcc test_nvcc.cu -o test_nvcc -run

CPU says hello.
GPU says hello.


## Pretrained Network를 로드해서 이미지를 샘플링 해봅시다.

혹시 만약에 

> tensorflow.python.framework.errors_impl.NotFoundError: /root/stylegan2_train/dnnlib/tflib/_cudacache/fused_bias_act_ec21d79f0dc288505704f796449a968e.so: undefined symbol: _ZN10tensorflow12OpDefBuilder6OutputESs


이 에러 로그가 발생할 경우, 
```stylegan2/dnnlib/tflib/custom_ops.py```로 가셔서 127번째 줄

```compile_opts += ' --compiler-options \'-fPIC -D_GLIBCXX_USE_CXX11_ABI=0\''``` 을 
```compile_opts += ' --compiler-options \'-fPIC -D_GLIBCXX_USE_CXX11_ABI=1\''``` 로 수정해주세요.


---

샘플링 실행 코드는 아래와 같습니다.

* --network : 네트워크의 pkl 파일이 담긴 경로 주소 (공식에서 제공합니다.)
* --seeds : code에서는 각 seed에서 하나씩 이미지를 샘플링 하는 것 같네요.
(예를 들면, 내가 이미지 10장을 뽑고 싶다면 1000-1009 로 설정 할 수 있는 것 같네요. 이건 공식 홈페이지에서도 참고할 수 있을 듯 합니다.)

* --truncation-psi : Truncation Trick을 통해 이미지의 다양성을 조절하면서 샘플링이 가능한 것 같습니다.
0일 수록 다양성이 없으며, 숫자가 커질 수록 다양성이 늘어납니다.


```!python run_generator.py generate-images --network=gdrive:networks/stylegan2-cat-config-f.pkl --seeds=1001-1004 --truncation-psi=0.1```



In [ ]:
# Generate uncurated cat images (matches paper Figure 12)
!python run_generator.py generate-images --network=gdrive:networks/stylegan2-cat-config-f.pkl --seeds=1000-5999 --truncation-psi=0

# (기타) NVIDIA CUDA 버전 및 CUDA Driver 버전 확인하기
이것은 Colab 사용 시 각자가 사용하는 GPU의 정보를 불러오는 코드입니다.

In [ ]:
!nvcc -V

In [ ]:
!nvidia-smi

# (진행중) LSUN-object Real Dataset Load

In [ ]:
!python lsun 

In [3]:
import tensorflow_datasets as tfds

In [ ]:
!python3 lsun/download_objects.py -c cat -o data